File: Week5/1 - Binary Tree Maximum Path Sum/Solution.cpp
/*
 * Created on Thu Apr 30 2020
 *
 * Title: Leetcode - Binary Tree Maximum Path Sum
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

#include <iostream>

using namespace std;

/**
 * Definition for a binary tree node as given below.
 * 
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int max_path_sum = 0;
    int dfs(TreeNode* root) {
        if(root == NULL) {
            return 0;
        }
        int left_sum = dfs(root->left);
        int right_sum = dfs(root->right);
        max_path_sum = max(max_path_sum, (root->val + left_sum + right_sum));
        return max(0, (root->val + max(left_sum, right_sum)));
    }

    int maxPathSum(TreeNode* root) {
        max_path_sum = INT32_MIN;
        dfs(root);
        return max_path_sum;
    }
};

File: Week5/1 - Binary Tree Maximum Path Sum/Solution.py
#
# Created on Thu Apr 30 2020
#
# Title: Leetcode - Binary Tree Maximum Path Sum
#
# Author: Vatsal Mistry
# Web: mistryvatsal.github.io
#

from sys import maxsize

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def dfs(self, root):
        if root is None:
            return 0
        left_sum, right_sum = self.dfs(root.left), self.dfs(root.right)
        self.max_path_sum = max(self.max_path_sum, root.val + left_sum + right_sum)
        return max(0, root.val+max(left_sum, right_sum))

    def maxPathSum(self, root):
        self.max_path_sum = -maxsize
        self.dfs(root)
        return self.max_path_sum
